File: 2130-max-twin-sum-linked-list/solution.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverse(head *ListNode) *ListNode {
	var prev *ListNode
	for head != nil {
		next := head.Next
		head.Next = prev
		prev = head
		head = next
	}
	return prev
}

func pairSum(head *ListNode) int {
	slow := head
	fast := head
	for fast != nil && fast.Next != nil && fast.Next.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	// slow var contains the middle node
	rev_head := reverse(slow)
	ans := 0
	for head != nil && rev_head != nil {
		ans = max(ans, head.Val+rev_head.Val)
		head = head.Next
		rev_head = rev_head.Next
	}
	return ans
}

File: 2130-max-twin-sum-linked-list/solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def reverse(self, head: ListNode) -> ListNode:
        prev = None
        while head:
            next = head.next
            head.next = prev
            prev = head
            head = next
        return prev

    def pairSum(self, head: ListNode) -> int:
        fast = head
        slow = head
        while fast and fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        
        rev_head = self.reverse(slow)
        ans = 0
        while head and rev_head:
            ans = max(ans, head.val + rev_head.val)
            head = head.next
            rev_head = rev_head.next
        return ans